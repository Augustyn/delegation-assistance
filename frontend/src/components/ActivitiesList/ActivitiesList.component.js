import React from "react";
import { array, func, bool } from "prop-types";
import Checkbox from "../Checkbox/Checkbox.component";
import Spinner from "../Spinner/Spinner.component";
import BlurredSpinnerWrapper from "../BlurredSpinnedWrapper/BlurredSpinnerWrapper.component";

import styles from "./ActivitiesList.module.scss";

const ActivitiesList = props => {
  const { activities = [], handleCheck, loading, updating } = props;
  return loading ? (
    <Spinner />
  ) : (
    <BlurredSpinnerWrapper loading={updating} message="saving checklist">
      {activities.map(activity => {
        return (
          <div key={activity.id} className="d-flex flex-column">
            <div className={styles.grow}>
              <Checkbox
                onClick={e => handleCheck(activity, e)}
                defaultChecked={activity.isDone}
                name={activity.id.toString()}
                disabled={updating}
              />
              <label htmlFor={activity.id.toString()}>{activity.task}</label>
              <p className={activity.isDone ? "strikethrough-text" : ""}>{activity.description}</p>
            </div>
          </div>
        );
      })}
    </BlurredSpinnerWrapper>
  );
};

ActivitiesList.propTypes = {
  activities: array,
  handleCheck: func,
  loading: bool,
  updating: bool
};

export default ActivitiesList;
