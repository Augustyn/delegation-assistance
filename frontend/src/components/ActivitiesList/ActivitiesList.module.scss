.grow {
  box-shadow: 0 0 1px rgba(0, 0, 0, 0);
  -webkit-transition-duration: 0.3s;
  transition-duration: 0.3s;
  -webkit-transition-property: transform;
  transition-property: transform;
  &:hover,
  &:focus,
  &:active {
    -webkit-transform: scale(1.02);
    transform: scale(1.02);
  }
}
