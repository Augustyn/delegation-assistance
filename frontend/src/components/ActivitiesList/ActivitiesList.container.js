import React, { Component } from "react";
import { connect } from "react-redux";
import { bool, func, object, number } from "prop-types";
import cloneDeep from "lodash/cloneDeep";
import debounce from "lodash/debounce";

import {
  getDelegationChecklist,
  getDelegationChecklistFetching,
  getDelegationChecklistUpdating
} from "../../selectors/delegationChecklist.selectors.js";
import { fetchDelegationChecklist, updateDelegationChecklist } from "../../actions/delegationChecklist.action";
import ActivitiesList from "./ActivitiesList.component.js";

export class ActivitiesListContainer extends Component {
  static propTypes = {
    delegationChecklist: object,
    delegationId: number,
    fetchDelegationChecklist: func,
    fetching: bool,
    updateDelegationChecklist: func,
    updating: bool
  };

  state = {
    delegationChecklist: {
      activities: []
    }
  };

  componentDidMount() {
    this.props.fetchDelegationChecklist(this.props.delegationId).then(() => {
      this.setState({ delegationChecklist: this.props.delegationChecklist });
    });
  }

  _debouceUpdateChecklist = debounce(
    activities =>
      this.props
        .updateDelegationChecklist(this.props.delegationId, activities)
        .finally(() => this.setState({ delegationChecklist: this.props.delegationChecklist })),
    800
  );

  _handleCheck = activitiy => {
    const activitiesCopy = cloneDeep(this.state.delegationChecklist.activities);
    activitiesCopy.find(ac => ac.id === activitiy.id).isDone = !activitiy.isDone;
    const checklistCopy = { activities: activitiesCopy, version: this.state.delegationChecklist.version };
    this.setState({ delegationChecklist: checklistCopy }, () =>
      this._debouceUpdateChecklist(this.state.delegationChecklist)
    );
  };

  render() {
    return (
      <ActivitiesList
        activities={this.state.delegationChecklist.activities}
        handleCheck={this._handleCheck}
        loading={this.props.fetching}
        updating={this.props.updating}
      />
    );
  }
}

const mapStateToProps = state => {
  return {
    delegationChecklist: getDelegationChecklist(state),
    fetching: getDelegationChecklistFetching(state),
    updating: getDelegationChecklistUpdating(state)
  };
};

const mapDispatchToProps = {
  fetchDelegationChecklist,
  updateDelegationChecklist
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(ActivitiesListContainer);
