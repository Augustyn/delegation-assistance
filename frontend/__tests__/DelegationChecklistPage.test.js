import React from "react";
import Enzyme, { shallow } from "enzyme";
import Adapter from "enzyme-adapter-react-16";

import ActivitiesList from "../src/components/ActivitiesList/ActivitiesList.component";
import mockDelegationChecklist from "../__mocks__/delegationChecklist.mock";

Enzyme.configure({ adapter: new Adapter() });

const setup = () => {
  const props = {
    handleCheck: jest.fn(),
    delegationID: "1",
    activities: mockDelegationChecklist,
    invalid: false
  };

  const enzymeWrapper = shallow(<ActivitiesList {...props} />);

  return {
    props,
    enzymeWrapper
  };
};

describe("Delegation checklist page", () => {
  let enzymeWrapper;
  let props;

  beforeEach(() => {
    ({ enzymeWrapper, props } = setup());
  });

  it("should render correctly", () => {
    expect(enzymeWrapper).toMatchSnapshot();
  });

  it("should render 2 Checkbox", () => {
    expect(enzymeWrapper.find("Checkbox").length).toBe(2);
  });

  it("should render 2 checkboxes labels", () => {
    expect(enzymeWrapper.find("label").length).toBe(2);
    expect(enzymeWrapper.find('label[htmlFor="0"]').text()).toBe("Buy bread");
    expect(enzymeWrapper.find('label[htmlFor="1"]').text()).toBe("Sell PC");
  });

  it("should call handlecheck", () => {
    enzymeWrapper
      .find('Checkbox[name="0"]')
      .dive()
      .simulate("change");
    expect(props.handleCheck).toBeCalled();
  });

  it("should render 2 checkboxes descriptions", () => {
    expect(enzymeWrapper.find("p").length).toBe(2);
  });
});
