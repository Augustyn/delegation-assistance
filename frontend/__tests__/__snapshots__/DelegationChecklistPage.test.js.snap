// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Delegation checklist page should render correctly 1`] = `
<BlurredSpinnerWrapper
  loading={true}
  message="saving checklist"
>
  <div
    className="d-flex flex-column"
    key="0"
  >
    <div>
      <Checkbox
        defaultChecked={false}
        name="0"
        onClick={[Function]}
      />
      <label
        htmlFor="0"
      >
        Buy bread
      </label>
      <p
        className=""
      >
        Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book. It has survived not only five centuries, but also the leap into electronic typesetting, remaining essentially unchanged. It was popularised in the 1960s with the release of Letraset sheets containing Lorem Ipsum passages, and more recently with desktop publishing software like Aldus PageMaker including versions of Lorem Ipsum.
      </p>
    </div>
  </div>
  <div
    className="d-flex flex-column"
    key="1"
  >
    <div>
      <Checkbox
        defaultChecked={true}
        name="1"
        onClick={[Function]}
      />
      <label
        htmlFor="1"
      >
        Sell PC
      </label>
      <p
        className="strikethrough-text"
      >
        desc2
      </p>
    </div>
  </div>
</BlurredSpinnerWrapper>
`;
